internal/provider/destination_resource.go

```diff
@@ -124,34 +141,26 @@ func (r *DestinationResource) Create(ctx context.Context, req resource.CreateReq
 		return
 	}
 
-	// Translate API response into Terraform model and save it into state
-	diagnostics := resp.State.Set(ctx, tfmodels.DestinationFromAPIModel(destination))
-	resp.Diagnostics.Append(diagnostics...)
+	r.SetStateData(ctx, &resp.State, resp.Diagnostics, destination)
 }
 
 func (r *DestinationResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
-	// Read Terraform prior state data into the model
-	var stateData tfmodels.Destination
-	resp.Diagnostics.Append(req.State.Get(ctx, &stateData)...)
+	destinationUUID := r.GetUUIDFromState(ctx, req.State, resp.Diagnostics)
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	destinationResp, err := r.client.Destinations().Get(ctx, stateData.UUID.ValueString())
+	destination, err := r.client.Destinations().Get(ctx, destinationUUID)
 	if err != nil {
 		resp.Diagnostics.AddError("Unable to Read Destination", err.Error())
 		return
 	}
 
-	// Translate API response into Terraform model and save it into state
-	diagnostics := resp.State.Set(ctx, tfmodels.DestinationFromAPIModel(destinationResp))
-	resp.Diagnostics.Append(diagnostics...)
+	r.SetStateData(ctx, &resp.State, resp.Diagnostics, destination)
 }
 
 func (r *DestinationResource) Update(ctx context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse) {
-	// Read Terraform plan data into the model
-	var planData tfmodels.Destination
-	resp.Diagnostics.Append(req.Plan.Get(ctx, &planData)...)
+	planData := r.GetPlanData(ctx, req.Plan, resp.Diagnostics)
```

Comment:
Maybe `r.GetPlanData()` should also return `Diagnostics.HasError()`.

Reply:
oh I like that